tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/IcuTextWriter.java

```diff
@@ -0,0 +1,311 @@
+// © 2019 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+package org.unicode.icu.tool.cldrtoicu;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static java.util.stream.Collectors.joining;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.Writer;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Writes an IcuData object to a text file. A lot of this class was copied directly from the
+ * original {@code IcuTextWriter} in the CLDR project and has a number of very idiosyncratic
+ * behaviours. The behaviour of this class is currently tuned to produce perfect parity with
+ * the original conversion tools, but once migration of the tools is complete, it should
+ * probably be revisited and tidied up.
+ */
+// TODO: Link to a definitive specification for the ICU data files and remove the hacks!
+final class IcuTextWriter {
+    private static final String INDENT = "    ";
+    // List of characters to escape in UnicodeSets
+    // ('\' followed by any of '\', '[', ']', '{', '}', '-', '&', ':', '^', '=').
+    private static final Pattern UNICODESET_ESCAPE =
+        Pattern.compile("\\\\[\\\\\\[\\]\\{\\}\\-&:^=]");
+    // Only escape \ and " from other strings.
+    private static final Pattern STRING_ESCAPE = Pattern.compile("(?!')\\\\\\\\(?!')");
+    private static final Pattern QUOTE_ESCAPE = Pattern.compile("\\\\?\"");
+
+    /** Write a file in ICU data format with the specified header. */
+    static void writeToFile(IcuData icuData, Path outDir, List<String> header) {
+        try {
+            Files.createDirectories(outDir);
+            try (Writer w = Files.newBufferedWriter(outDir.resolve(icuData.getName() + ".txt"));
+                PrintWriter out = new PrintWriter(w)) {
+                new IcuTextWriter(icuData).writeTo(out, header);
+            }
+        } catch (IOException e) {
+            throw new RuntimeException("cannot write ICU data file: " + icuData.getName(), e);
+        }
+    }
+
+    private final IcuData icuData;
+    private int depth = 0;
+    private boolean valueWasInline = false;
+
+    IcuTextWriter(IcuData icuData) {
+        this.icuData = checkNotNull(icuData);
+    }
+
+    private void writeTo(PrintWriter out, List<String> header) throws IOException {
+        out.write('\uFEFF');
```

Comment:
related to https://unicode-org.atlassian.net/browse/ICU-10197 - may want to add the utf-8 coding to the header. (Can be a follow up).

Reply:
Ta, I'll add a TODO now and I (or anyone else) can followup later.